File: go/db2.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"time"
)

type Auth struct {
	Token string `json:"token"`
}

type AuthSettings struct {
	RestHttps bool
	Hostname  string
	RestPort  int
	Database  string
	DbPort    int
	SslDb2    bool
	Password  string
	Username  string
	Expiry    string
}

type DbParms struct {
	DbHost          string `json:"dbHost"`
	DbName          string `json:"dbName"`
	DbPort          int    `json:"dbPort"`
	IsSSLConnection bool   `json:"isSSLConnection"`
	Password        string `json:"password"`
	Username        string `json:"username"`
}

type AuthBody struct {
	DbParms    DbParms `json:"dbParms"`
	ExpiryTime string  `json:"expiryTime"`
}

type QueryParams[T any] struct {
	Parameters T    `json:"parameters"`
	Sync       bool `json:"sync"`
}

type SQLParams[T any] struct {
	SQLStatement string `json:"sqlStatement"`
	Parameters   T      `json:"parameters"`
	IsQuery      bool   `json:"isQuery"`
	Sync         bool   `json:"sync"`
}

type Status int

const (
	Failed Status = iota
	New
	Running
	DataAvailable
	Completed
	Stopping
)

type RawResponse[T any] struct {
	JobStatus int `json:"jobStatus"`
	ResultSet []T `json:"resultSet"`
}

type Response[T any] struct {
	Status  Status
	Results []T
}

type JobResponse struct {
	Id string `json:"id"`
}

type Job[T any] struct {
	NextPageUrl string
	StopUrl     string
	authToken   string
}

type Db2REST struct {
	AuthSettings AuthSettings
	authToken    string
}

func GetDb2AuthToken(authSettings AuthSettings) (string, error) {
	method := "https"
	if !authSettings.RestHttps {
		method = "http"
	}
	authUrl := fmt.Sprintf("%s://%s:%d/v1/auth",
		method, authSettings.Hostname, authSettings.RestPort)

	body := AuthBody{
		DbParms: DbParms{
			DbHost:          authSettings.Hostname,
			DbName:          authSettings.Database,
			DbPort:          authSettings.DbPort,
			IsSSLConnection: authSettings.SslDb2,
			Password:        authSettings.Password,
			Username:        authSettings.Username,
		},
		ExpiryTime: authSettings.Expiry,
	}

	client := &http.Client{}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", authUrl, bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	defer resp.Body.Close()
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode == 200 {
		var auth Auth
		err = json.Unmarshal(respBody, &auth)
		if err != nil {
			return "", err
		}

		return auth.Token, nil
	} else {
		return "", errors.New(string(respBody))
	}
}

func toStatus(n int) Status {
	switch n {
	case 0:
		return Failed
	case 1:
		return New
	case 2:
		return Running
	case 3:
		return DataAvailable
	case 4:
		return Completed
	case 5:
		return Stopping
	default:
		panic("unreachable")
	}
}

func toResponse[T any](raw RawResponse[T]) Response[T] {
	return Response[T]{
		Status:  toStatus(raw.JobStatus),
		Results: raw.ResultSet,
	}
}

func Connect(settings AuthSettings) (Db2REST, error) {
	token, err := GetDb2AuthToken(settings)
	if err != nil {
		return Db2REST{}, err
	}

	return Db2REST{
		AuthSettings: settings,
		authToken:    token,
	}, nil
}

func RunService[T any](db2 *Db2REST, service string, version string, parameters T, sync bool) (string, error) {
	method := "https"
	if !db2.AuthSettings.RestHttps {
		method = "http"
	}
	url := fmt.Sprintf("%s://%s:%d/v1/services/%s/%s",
		method, db2.AuthSettings.Hostname, db2.AuthSettings.RestPort,
		service, version)

	body := QueryParams[T]{
		Parameters: parameters,
		Sync:       sync,
	}

	client := &http.Client{}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("authorization", db2.authToken)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	defer resp.Body.Close()
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode == 200 || resp.StatusCode == 202 {
		return string(respBody), nil
	} else {
		return "", errors.New(string(respBody))
	}
}

func RunSQL[T any](db2 *Db2REST, statement string, parameters T, isQuery bool, sync bool) (string, error) {
	method := "https"
	if !db2.AuthSettings.RestHttps {
		method = "http"
	}
	url := fmt.Sprintf("%s://%s:%d/v1/services/execsql",
		method, db2.AuthSettings.Hostname, db2.AuthSettings.RestPort)

	body := SQLParams[T]{
		SQLStatement: statement,
		Parameters:   parameters,
		IsQuery:      isQuery,
		Sync:         sync,
	}

	client := &http.Client{}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("authorization", db2.authToken)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	defer resp.Body.Close()
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode == 200 || resp.StatusCode == 202 {
		return string(respBody), nil
	} else {
		return "", errors.New(string(respBody))
	}
}

func RunSyncQueryService[T any, U any](db2 *Db2REST, service string, version string, parameters T) (Response[U], error) {
	response, err := RunService[T](db2, service, version, parameters, true)
	if err != nil {
		return Response[U]{}, err
	}

	var raw RawResponse[U]
	err = json.Unmarshal([]byte(response), &raw)
	if err != nil {
		return Response[U]{}, err
	}

	return toResponse[U](raw), nil
}

func RunSyncQuerySQL[T any, U any](db2 *Db2REST, statement string, parameters T) (Response[U], error) {
	response, err := RunSQL[T](db2, statement, parameters, true, true)
	if err != nil {
		return Response[U]{}, err
	}

	var raw RawResponse[U]
	err = json.Unmarshal([]byte(response), &raw)
	if err != nil {
		return Response[U]{}, err
	}

	return toResponse[U](raw), nil
}

func RunSyncStatementService[T any](db2 *Db2REST, service string, version string, parameters T) error {
	_, err := RunService[T](db2, service, version, parameters, true)
	if err != nil {
		return err
	}

	return nil
}

func RunSyncStatementSQL[T any](db2 *Db2REST, statement string, parameters T) error {
	_, err := RunSQL[T](db2, statement, parameters, false, true)
	if err != nil {
		return err
	}

	return nil
}

func RunAsyncService[T any, U any](db2 *Db2REST, service string, version string, parameters T) (Job[U], error) {
	response, err := RunService[T](db2, service, version, parameters, false)
	if err != nil {
		return Job[U]{}, err
	}

	var jobId JobResponse
	err = json.Unmarshal([]byte(response), &jobId)
	if err != nil {
		return Job[U]{}, err
	}

	return NewJob[U](jobId.Id, &db2.AuthSettings, db2.authToken), nil
}

func RunAsyncSQL[T any, U any](db2 *Db2REST, statement string, parameters T) (Job[U], error) {
	response, err := RunSQL[T](db2, statement, parameters, true, false)
	if err != nil {
		return Job[U]{}, err
	}

	var jobId JobResponse
	err = json.Unmarshal([]byte(response), &jobId)
	if err != nil {
		return Job[U]{}, err
	}

	return NewJob[U](jobId.Id, &db2.AuthSettings, db2.authToken), nil
}

func NewJob[T any](jobId string, authSettings *AuthSettings, authToken string) Job[T] {
	method := "https"
	if !authSettings.RestHttps {
		method = "http"
	}
	nextPageUrl := fmt.Sprintf("%s://%s:%d/v1/services/%s",
		method, authSettings.Hostname, authSettings.RestPort, jobId)
	stopUrl := fmt.Sprintf("%s://%s:%d/v1/services/stop/%s",
		method, authSettings.Hostname, authSettings.RestPort, jobId)

	return Job[T]{
		NextPageUrl: nextPageUrl,
		StopUrl:     stopUrl,
		authToken:   authToken,
	}
}

func (job *Job[T]) StopJob() error {
	client := &http.Client{}
	req, err := http.NewRequest("GET", job.StopUrl, nil)
	if err != nil {
		return err
	}

	req.Header.Set("authorization", job.authToken)
	resp, err := client.Do(req)
	if err != nil {
		return err
	}

	defer resp.Body.Close()
	if resp.StatusCode == 204 {
		return nil
	} else {
		respBody, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			return err
		}

		return errors.New(string(respBody))
	}
}

func (job *Job[T]) CurrentPage(limit int) (*Response[T], error) {
	limithm := map[string]int{
		"limit": limit,
	}

	client := &http.Client{}
	jsonBody, err := json.Marshal(limithm)
	if err != nil {
		return &Response[T]{}, err
	}

	req, err := http.NewRequest("GET", job.NextPageUrl, bytes.NewBuffer(jsonBody))
	if err != nil {
		return &Response[T]{}, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("authorization", job.authToken)
	resp, err := client.Do(req)
	if err != nil {
		return &Response[T]{}, err
	}

	defer resp.Body.Close()
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return &Response[T]{}, err
	}

	if resp.StatusCode == 404 {
		return nil, nil
	} else if resp.StatusCode == 200 {
		var raw RawResponse[T]
		err = json.Unmarshal(respBody, &raw)
		if err != nil {
			return &Response[T]{}, err
		}

		r := toResponse[T](raw)
		return &r, nil
	} else {
		return &Response[T]{}, errors.New(string(respBody))
	}
}

func (job *Job[T]) NextPage(limit int, refresh time.Duration) (*Response[T], error) {
	for {
		page, err := job.CurrentPage(limit)
		if err != nil {
			return &Response[T]{}, err
		}

		if page == nil {
			return nil, nil
		}

		if page.Status == New || page.Status == Running {
			<-time.After(refresh)
			continue
		}

		return page, nil
	}
}

File: swift/db2.swift
import Foundation

public class Db2REST {
    public struct AuthSettings: Codable {
        var hostname: String
        var database: String
        var dbPort: Int
        var restPort: Int
        var sslDb2: Bool
        var sslRest: Bool
        var password: String
        var username: String
        var expiryTime: String
        
        var urlProtocol: String {
            sslRest ? "https" : "http"
        }

        public init(hostname: String, database: String, dbPort: Int,
                    restPort: Int, sslDb2: Bool, sslRest: Bool,
                    password: String, username: String, expiryTime: String) {
            self.hostname = hostname
            self.database = database
            self.dbPort = dbPort
            self.restPort = restPort
            self.sslDb2 = sslDb2
            self.sslRest = sslRest
            self.password = password
            self.username = username
            self.expiryTime = expiryTime
        }
    }
    
    private struct Auth: Codable {
        var token: String
    }
    
    private struct QueryResponse<T: Codable>: Codable {
        var jobStatus: Int
        var jobStatusDescription: String?
        var resultSet: [T]?
        var rowCount: Int
        
        func export() -> Response<T> {
            return Response(status: Status.from(jobStatus), results: resultSet)
        }
    }
    
    public enum Status {
        case failed
        case new
        case running
        case dataAvailable
        case completed
        case stopping
        
        static func from(_ n: Int) -> Self {
            return [.failed, .new, .running, .dataAvailable, .completed, .stopping][n]
        }
    }
    
    public struct Response<T> {
        var status: Status
        var results: [T]?
    }
    
    private struct JobResponse: Codable {
        var id: String
    }
    
    public actor Job<T: Codable> {
        enum JobError: Error {
            case failure(String?)
            case cancelled
        }
        
        private var executing = false
        private var stopped = false
        private var pageRequest: URLRequest!
        private var stopRequest: URLRequest!
        
        deinit {
            if !stopped {
                let request = stopRequest
                Task {
                    try? await Job.stopJob(stopRequest: request!)
                }
            }
        }
        
        private static func stopJob(stopRequest: URLRequest) async throws {
            let (result, response) = try await URLSession.shared.data(for: stopRequest)
            guard (response as? HTTPURLResponse)?.statusCode == 204 else {
                throw RequestError.invalidResponse(String(data: result, encoding: .utf8)!)
            }
        }
       
        init(jobID: String, authSettings: AuthSettings, authToken: String) async throws {
            guard let nextPageURL = URL(string: "\(authSettings.urlProtocol)://\(authSettings.hostname):\(authSettings.restPort)/v1/services/\(jobID)") else {
                throw RequestError.invalidURL
            }
            pageRequest = URLRequest(url: nextPageURL, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 5)
            pageRequest.httpMethod = "POST"
            pageRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
            pageRequest.addValue(authToken, forHTTPHeaderField: "authorization")
            
            guard let stopURL = URL(string: "http://\(authSettings.hostname):\(authSettings.restPort)/v1/services/stop/\(jobID)") else {
                throw RequestError.invalidURL
            }
            stopRequest = URLRequest(url: stopURL, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 5)
            stopRequest.httpMethod = "PUT"
            stopRequest.addValue(authToken, forHTTPHeaderField: "authorization")
        }
        
        private func nextRawPage(limit: Int) async throws -> QueryResponse<T>? {
            pageRequest.httpBody = try JSONSerialization.data(withJSONObject: ["limit": limit], options: .init())
            let (result, response) = try await URLSession.shared.data(for: pageRequest)
            let statusCode = (response as? HTTPURLResponse)?.statusCode
            guard statusCode == 200 else {
                if statusCode == 404 {
                    return nil
                }
                throw RequestError.invalidResponse(String(data: result, encoding: .utf8)!)
            }
            return try JSONDecoder().decode(QueryResponse<T>.self, from: result)
        }
        
        private func stopJob() async throws {
            defer {
                stopped = true
            }
            try await Job.stopJob(stopRequest: stopRequest)
        }
        
        func nextPage(limit: Int) async throws -> Response<T>? {
            if executing {
                return nil
            }
            executing = true
            defer {
                executing = false
            }
            do {
                while let page = try await nextRawPage(limit: limit) {
                    if page.jobStatus == 0 {
                        return Response(status: .failed, results: nil)
                    }
                    if page.jobStatus == 1 || page.jobStatus == 2 {
                        try await Task.sleep(nanoseconds: 1_000_000_000)
                        continue
                    }
                    return page.export()
                }
            } catch let error {
                Task {
                    try await stopJob()
                }
                throw error
            }
            return nil
        }
    }

    enum AuthenticationError: Error {
        case noAuthToken
    }
    
    enum RequestError: Error {
        case invalidURL
        case invalidResponse(String)
    }

    public let authSettings: AuthSettings
    private var authToken: String!
    private var authTokenError: Error!
    
    public init(authSettings: AuthSettings) async throws {
        self.authSettings = authSettings
        authToken = try await getDb2AuthToken()
        if authToken == nil {
            throw AuthenticationError.noAuthToken
        }
    }
    
    private func getDb2AuthToken() async throws -> String {
        guard let authUrl = URL(string: "\(authSettings.urlProtocol)://\(authSettings.hostname):\(authSettings.restPort)/v1/auth") else {
            throw RequestError.invalidURL
        }

        let body: [String: Any] = [
            "dbParms": [
                "dbHost": authSettings.hostname,
                "dbName": authSettings.database,
                "dbPort": authSettings.dbPort,
                "isSSLConnection": authSettings.sslDb2,
                "password": authSettings.password,
                "username": authSettings.username
            ],
            "expiryTime": authSettings.expiryTime
        ]
        
        print(String(data: try! JSONSerialization.data(withJSONObject: body, options: .prettyPrinted), encoding: .utf8)!)

        var request = URLRequest(url: authUrl, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .init())
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let (result, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else {
            print(String(data: result, encoding: .utf8)!)
            throw RequestError.invalidResponse(String(data: result, encoding: .utf8)!)
        }
        return try JSONDecoder().decode(Auth.self, from: result).token
    }
    
    private func runQuery(service: String, version: String, parameters: [String: Any], sync: Bool) async throws -> Data {
        guard let uploadUrl = URL(string: "\(authSettings.urlProtocol)://\(authSettings.hostname):\(authSettings.restPort)/v1/services/\(service)/\(version)") else {
            throw RequestError.invalidURL
        }

        let body: [String: Any] = [
            "parameters": parameters,
            "sync": sync
        ]

        var request = URLRequest(url: uploadUrl, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .init())
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue(authToken, forHTTPHeaderField: "authorization")
        
        let (result, response) = try await URLSession.shared.data(for: request)
        try Task.checkCancellation()
        let statusCode = (response as? HTTPURLResponse)?.statusCode
        guard statusCode == 200 || statusCode == 202 else { throw RequestError.invalidResponse(String(data: result, encoding: .utf8)!) }
        return result
    }

    private func runSQL(statement: String, parameters: [String: Any], isQuery: Bool, sync: Bool) async throws -> Data {
        guard let url = URL(string: "\(authSettings.urlProtocol)://\(authSettings.hostname):\(authSettings.restPort)/v1/services/execsql") else {
            throw RequestError.invalidURL
        }

        let body: [String: Any] = [
            "sqlStatement": statement,
            "parameters": parameters,
            "isQuery": isQuery,
            "sync": sync
        ]

        var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .init())
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue(authToken, forHTTPHeaderField: "authorization")

        let (result, response) = try await URLSession.shared.data(for: request)
        try Task.checkCancellation()
        let statusCode = (response as? HTTPURLResponse)?.statusCode
        guard statusCode == 200 || statusCode == 202 else { throw RequestError.invalidResponse(String(data: result, encoding: .utf8)!) }
        return result
    }
    
    func runSyncJob<T: Codable>(service: String, version: String, parameters: [String: Any]) async throws -> Response<T>? {
        let result = try await runQuery(service: service, version: version, parameters: parameters, sync: true)
        return try JSONDecoder().decode(QueryResponse<T>.self, from: result).export()
    }

    func runSyncSQL<T: Codable>(statement: String, parameters: [String: Any]) async throws -> Response<T>? {
        let result = try await runSQL(statement: statement, parameters: parameters, isQuery: true, sync: true)
        return try JSONDecoder().decode(QueryResponse<T>.self, from: result).export()
    }
    
    func runSyncJob(service: String, version: String, parameters: [String: Any]) async throws {
        _ = try await runQuery(service: service, version: version, parameters: parameters, sync: true)
    }

    func runSyncSQL(statement: String, parameters: [String: Any]) async throws {
        _ = try await runSQL(statement: statement, parameters: parameters, isQuery: false, sync: true)
    }
    
    func runAsyncJob<T>(service: String, version: String, parameters: [String: Any]) async throws -> Job<T> {
        let result = try await runQuery(service: service, version: version, parameters: parameters, sync: false)
        let jobId = try JSONDecoder().decode(JobResponse.self, from: result).id
        return try await Job<T>(jobID: jobId, authSettings: authSettings, authToken: authToken)
    }

    func runAsyncSQL<T>(statement: String, parameters: [String: Any]) async throws -> Job<T> {
        let result = try await runSQL(statement: statement, parameters: parameters, isQuery: true, sync: false)
        let jobId = try JSONDecoder().decode(JobResponse.self, from: result).id
        return try await Job(jobID: jobId, authSettings: authSettings, authToken: authToken)
    }
}

protocol Swiftifiable {
    associatedtype SwiftifiedType
    
    func convert() -> SwiftifiedType
}

extension Optional: Swiftifiable where Wrapped: Swiftifiable {
    func convert() -> Wrapped.SwiftifiedType? {
        self == nil ? nil : self.convert()
    }
}

extension Db2REST.Response: Swiftifiable where T: Swiftifiable {
    func convert() -> Db2REST.Response<T.SwiftifiedType> {
        Db2REST.Response(status: status, results: results == nil ? nil : results!.map { $0.convert() })
    }
}

extension Db2REST.Job: Swiftifiable where T: Swiftifiable {
    nonisolated func convert() -> SwiftifiedJob<T> {
        SwiftifiedJob(job: self)
    }
}

actor SwiftifiedJob<T: Codable> where T: Swiftifiable {
    private let job: Db2REST.Job<T>
    
    init(job: Db2REST.Job<T>) {
        self.job = job
    }
    
    func nextPage(limit: Int) async throws -> Db2REST.Response<T.SwiftifiedType>? {
        try await job.nextPage(limit: limit)?.convert()
    }
}

struct Db2NativeUtils {
    @inlinable
    static func convert(_ x: Int) -> Int32 {
        return Int32(x)
    }
    
    @inlinable
    static func convert(_ x: Int32) -> Int {
        return Int(x)
    }
    
    @inlinable
    static func convert(_ x: Int?) -> Int32? {
        return x == nil ? nil : Int32(x!)
    }
    
    @inlinable
    static func convert(_ x: Int32?) -> Int? {
        return x == nil ? nil : Int(x!)
    }
    
    @inlinable
    static func convert<T>(_ x: T) -> T {
        return x
    }
}

struct Nothing: Codable {}
